# Burke-Lab-SNP-calling-pipeline
Scripts for use with Pool-SEQ data to generate tables of SNP frequencies as txt files.


This pipeline was written for the Burke Lab at Oregon State U by Shawn O'Neill, Dana Gibbon, and Molly Burke.

This pipeline runs GATK 4.X on a set of populations/samples to produce two tables of parsed and filtered variant calls: filtered_snps.txt and 
filtered_indels.txt.

For users at Oregon State University, the run_pipeline.sh script (and updates, see below) can be customized (see the readme within) and used with SGE tools configured for OSU's CGRB research computing cluster.  

Users outside of the OSU environment should reference the scripts in local/bin/ for details of the filtering parameters used with GATK, as well as how the VCF files generated by GATK are converted into SNP tables in txt format.

v1 of the pipeline (used to generate yeast SNP tables prior to 2021)
run_pipeline.sh = requires SE data, creates a single merged vcf (and converts to a SNP table) but must include all raw sequence data to be included

v2 of the pipeline (used to generate yeast SNP tables post-2021)
run_pipeline-add-samples-01.sh = part 1 of an update that allows a 2-part SNP calling process which facilitates adding data to a project; creates individual vcf files for all samples
run_pipeline-add-samples-02.sh = part 2 of the update - run this after vcf files are produced for all samples of interest (this merges into a single vcf table, then converts to a SNP table)
run_pipeline-add-samples-01_paired.sh = update to run_pipeline-add-samples-01.sh that allows PE data
